File: app/src/main/java/com/njlabs/showjava/processor/JarExtractor.java
package com.njlabs.showjava.processor;

import android.util.Log;

import com.crashlytics.android.Crashlytics;
import com.googlecode.dex2jar.Method;
import com.googlecode.dex2jar.ir.IrMethod;
import com.googlecode.dex2jar.reader.DexFileReader;
import com.googlecode.dex2jar.v3.Dex2jar;
import com.googlecode.dex2jar.v3.DexExceptionHandler;
import com.njlabs.showjava.utils.StringUtils;
import com.njlabs.showjava.utils.logging.Ln;

import org.jf.dexlib2.DexFileFactory;
import org.jf.dexlib2.iface.ClassDef;
import org.jf.dexlib2.iface.DexFile;
import org.jf.dexlib2.immutable.ImmutableDexFile;
import org.objectweb.asm.tree.MethodNode;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintStream;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by Niranjan on 29-05-2015.
 */
@SuppressWarnings({"ResultOfMethodCallIgnored", "ConstantConditions"})
public class JarExtractor extends ProcessServiceHelper {

    private final ArrayList<String> ignoredLibs;

    public JarExtractor(ProcessService processService) {
        this.processService = processService;
        this.UIHandler = processService.UIHandler;
        this.packageFilePath = processService.packageFilePath;
        this.packageName = processService.packageName;
        this.exceptionHandler = processService.exceptionHandler;
        this.sourceOutputDir = processService.sourceOutputDir;
        this.javaSourceOutputDir = processService.javaSourceOutputDir;
        ignoredLibs = new ArrayList<>();

        //////
        printStream = new PrintStream(new ProgressStream());
        System.setErr(printStream);
        System.setOut(printStream);
        //////
    }

    public void extract() {
        ThreadGroup group = new ThreadGroup("DEX TO JAR EXTRACTION");
        broadcastStatus("optimise_dex_start");
        Runnable runProcess = new Runnable() {
            @Override
            public void run() {
                loadIgnoredLibs();
                apkToDex();
                if(!processService.decompilerToUse.equals("jadx")){
                    dexToJar();
                }
                startJavaExtractor();
            }
        };
        Thread extractionThread = new Thread(group, runProcess, "DEX TO JAR EXTRACTION", processService.STACK_SIZE);
        extractionThread.setPriority(Thread.MAX_PRIORITY);
        extractionThread.setUncaughtExceptionHandler(exceptionHandler);
        extractionThread.start();
    }

    private void apkToDex() {
        DexFile dexFile = null;
        try {
            dexFile = DexFileFactory.loadDexFile(packageFilePath, 19);
        } catch (Exception e) {
            broadcastStatus("exit");
            UIHandler.post(new ToastRunnable("The app you selected cannot be decompiled. Please select another app."));
        }
        List<ClassDef> classes = new ArrayList<>();
        broadcastStatus("optimising", "");

        for (ClassDef classDef : dexFile.getClasses()) {
            if (!isIgnored(classDef.getType())) {
                final String CurrentClass = classDef.getType();
                broadcastStatus("optimising_class", CurrentClass.replaceAll("Processing ", ""));
                classes.add(classDef);
            }
        }
        broadcastStatus("optimise_dex_finish");

        File PerAppWorkingDirectory = new File(processService.sourceOutputDir);
        PerAppWorkingDirectory.mkdirs();

        Log.d("DEBUGGER", "Prepare Writing");

        broadcastStatus("merging_classes");

        dexFile = new ImmutableDexFile(classes);

        try {
            Log.d("DEBUGGER", "Start Writing");
            DexFileFactory.writeDexFile(PerAppWorkingDirectory + "/optimised_classes.dex", dexFile);
            Log.d("DEBUGGER", "Writing done!");
        } catch (Exception e) {
            broadcastStatus("exit");
            UIHandler.post(new ToastRunnable("The app you selected cannot be decompiled. Please select another app."));
        }


    }

    private void dexToJar() {
        Log.i("STATUS", "Jar Extraction Started");

        broadcastStatus("dex2jar");

        // DEX 2 JAR CONFIGS
        boolean reuseReg = false; // reuse register while generate java .class file
        boolean topologicalSort1 = false; // same with --topological-sort/-ts
        boolean topologicalSort = false; // sort block by topological, that will generate more readable code
        boolean verbose = true; // show progress
        boolean debugInfo = false; // translate debug info
        boolean printIR = false; // print ir to System.out
        boolean optimizeSynchronized = true; // Optimise-synchronised

        File PerAppWorkingDirectory = new File(sourceOutputDir);
        File file = new File(PerAppWorkingDirectory + "/" + packageName + ".jar");

        File dexFile = new File(PerAppWorkingDirectory + "/optimised_classes.dex");

        if (dexFile.exists() && dexFile.isFile()) {
            DexExceptionHandlerMod dexExceptionHandlerMod = new DexExceptionHandlerMod();
            try {
                DexFileReader reader = new DexFileReader(dexFile);
                Dex2jar dex2jar = Dex2jar.from(reader).reUseReg(reuseReg).topoLogicalSort(topologicalSort || topologicalSort1).skipDebug(!debugInfo)
                        .optimizeSynchronized(optimizeSynchronized).printIR(printIR).verbose(verbose);
                dex2jar.setExceptionHandler(dexExceptionHandlerMod);
                dex2jar.to(file);
            } catch (Exception e) {
                broadcastStatus("exit_process_on_error");
            }

            Log.i("STATUS", "Clearing cache");
            File ClassDex = new File(PerAppWorkingDirectory + "/optimised_classes.dex");
            ClassDex.delete();
        }

    }

    private void startJavaExtractor() {
        JavaExtractor javaExtractor = new JavaExtractor(processService);
        javaExtractor.extract();
    }

    private void loadIgnoredLibs() {
        String ignoredList = (processService.IGNORE_LIBS ? "ignored.list":"ignored_basic.list");
        BufferedReader reader = null;
        try {
            reader = new BufferedReader(new InputStreamReader(processService.getAssets().open(ignoredList)));
            String mLine = reader.readLine().trim();
            while (mLine != null) {
                mLine = mLine.trim();
                if (mLine.length() != 0) {
                    ignoredLibs.add(StringUtils.toClassName(mLine));
                }
                mLine = reader.readLine();
            }
        } catch (IOException e) {
            Crashlytics.logException(e);
        } finally {
            if (reader != null) {
                try {
                    reader.close();
                } catch (IOException e) {
                    Crashlytics.logException(e);
                }
            }
        }
    }

    private boolean isIgnored(String className) {
        for (String ignoredClass : ignoredLibs) {
            if (className.startsWith(ignoredClass)) {
                return true;
            }
        }
        return false;
    }

    private class DexExceptionHandlerMod implements DexExceptionHandler {
        @Override
        public void handleFileException(Exception e) {
            Ln.d("Dex2Jar Exception " + e);
        }

        @Override
        public void handleMethodTranslateException(Method method, IrMethod irMethod, MethodNode methodNode, Exception e) {
            Ln.d("Dex2Jar Exception " + e);
        }
    }


}

File: app/src/main/kotlin/com/njlabs/showjava/services/processor/handlers/JarExtractor.kt
package com.njlabs.showjava.services.processor.handlers

import com.googlecode.dex2jar.Method
import com.googlecode.dex2jar.ir.IrMethod
import com.googlecode.dex2jar.reader.DexFileReader
import com.googlecode.dex2jar.v3.Dex2jar
import com.googlecode.dex2jar.v3.DexExceptionHandler
import com.njlabs.showjava.services.processor.ProcessorService
import com.njlabs.showjava.utils.StringTools
import org.jf.dexlib2.DexFileFactory
import org.jf.dexlib2.Opcodes
import org.jf.dexlib2.iface.ClassDef
import org.jf.dexlib2.iface.DexFile
import org.jf.dexlib2.immutable.ImmutableDexFile
import org.objectweb.asm.tree.MethodNode
import timber.log.Timber
import java.io.File
import java.io.PrintStream


class JarExtractor(private val processorService: ProcessorService) : BaseHandler(processorService) {

    private var ignoredLibs: ArrayList<String>? = ArrayList()
    private val perAppWorkingDirectory: File = File(processorService.sourceOutputDirectory)

    private var dexFile: File = File(perAppWorkingDirectory, "optimised_classes.dex")
    private var jarFile: File =
        File(perAppWorkingDirectory, "${processorService.inputPackageName}.jar")

    init {
        printStream = PrintStream(ProgressStream())
        System.setErr(printStream)
        System.setOut(printStream)
        Timber.tag("JarExtractor")
    }

    fun extract() {
        loadIgnoredLibs()
        convertApkToDex()
        if (processorService.decompilerToUse != "jadx") {
            convertDexToJar()
        }
    }

    private fun loadIgnoredLibs() {
        val ignoredList =
            if (processorService.shouldIgnoreLibs) "ignored.list" else "ignored_basic.list"
        processorService.application.assets.open(ignoredList).bufferedReader().useLines {
            it.map { line -> ignoredLibs?.add(StringTools.toClassName(line)) }
        }
    }

    private fun convertApkToDex() {

        Timber.i("Starting APK to DEX Conversion")

        var dexFile: DexFile? = null
        try {
            dexFile = DexFileFactory.loadDexFile("", Opcodes.getDefault())
        } catch (e: Exception) {
            broadcastStatus("exit")
        }

        val classes = ArrayList<ClassDef>()
        broadcastStatus("optimising")

        for (classDef in dexFile!!.classes) {
            if (!isIgnored(classDef.type)) {
                val currentClass = classDef.type
                broadcastStatus(
                    "optimising_class",
                    currentClass.replace("Processing ".toRegex(), "")
                )
                classes.add(classDef)
            }
        }

        broadcastStatus("optimise_dex_finish")
        Timber.i("Output directory: $perAppWorkingDirectory")
        broadcastStatus("merging_classes")
        dexFile = ImmutableDexFile(Opcodes.getDefault(), classes)

        try {
            this.dexFile = File(perAppWorkingDirectory, "optimised_classes.dex")
            DexFileFactory.writeDexFile(
                this.dexFile.absolutePath,
                dexFile
            )
            Timber.i("DEX file location: ${this.dexFile}")
        } catch (e: Exception) {
            Timber.e(e)
            broadcastStatus("exit", "cannot_decompile")
        }

    }

    private fun convertDexToJar() {
        Timber.i("Starting DEX to JAR Conversion")
        broadcastStatus("dex2jar")

        val reuseReg = false // reuse register while generate java .class file
        val topologicalSort1 = false // same with --topological-sort/-ts
        val topologicalSort =
            false // sort block by topological, that will generate more readable code
        val verbose = true // show progress
        val debugInfo = false // translate debug info
        val printIR = false // print ir to System.out
        val optimizeSynchronized = true // Optimise-synchronised

        if (dexFile.exists() && dexFile.isFile) {
            val dexExceptionHandlerMod = DexExceptionHandlerMod()
            try {
                val reader = DexFileReader(dexFile)
                val dex2jar = Dex2jar.from(reader).reUseReg(reuseReg)
                    .topoLogicalSort(topologicalSort || topologicalSort1).skipDebug(!debugInfo)
                    .optimizeSynchronized(optimizeSynchronized).printIR(printIR).verbose(verbose)
                dex2jar.exceptionHandler = dexExceptionHandlerMod
                dex2jar.to(jarFile)
            } catch (e: Exception) {
                broadcastStatus("exit_process_on_error")
            }
            Timber.i("Clearing cache")
            dexFile.delete()
        }
    }


    private fun isIgnored(className: String): Boolean {
        return ignoredLibs!!.any { className.startsWith(it) }
    }

    private inner class DexExceptionHandlerMod : DexExceptionHandler {
        override fun handleFileException(e: Exception) {
            Timber.d("Dex2Jar Exception $e")
        }

        override fun handleMethodTranslateException(
            method: Method,
            irMethod: IrMethod,
            methodNode: MethodNode,
            e: Exception
        ) {
            Timber.d("Dex2Jar Exception $e")
        }
    }

}